include/nbl/video/IGPURayTracingPipeline.h

```diff
@@ -15,6 +15,147 @@ class IGPURayTracingPipeline :  public IGPUPipeline<asset::IRayTracingPipeline<c
+                          // every shader must not be null. use SIndex::Unused to represent unused shader.
+                          return false;
+                      }
+                  }
+
+                  auto getShaderStage = [this](size_t index) -> hlsl::ShaderStage
+                  {
+                      return shaders[index].stage;
+                  };
+
+                auto isValidShaderIndex = [this, getShaderStage](size_t index, hlsl::ShaderStage expectedStage, bool is_unused_shader_forbidden) -> bool
+                  {
+                    if (index == SShaderGroupsParams::SIndex::Unused)
+                      return !is_unused_shader_forbidden;
+                    if (index >= shaders.size())
+                      return false;
+                    if (getShaderStage(index) != expectedStage)
+                      return false;
+                    return true;
+                  };
+
+                if (!isValidShaderIndex(shaderGroups.raygen.index, hlsl::ShaderStage::ESS_RAYGEN, true))
+                {
+                  return false;
+                }
+
+                for (const auto& shaderGroup : shaderGroups.hits)
+                {
+                  // https://docs.vulkan.org/spec/latest/chapters/pipelines.html#VUID-VkRayTracingPipelineCreateInfoKHR-flags-03470
+                  if (!isValidShaderIndex(shaderGroup.anyHit, 
+                    hlsl::ShaderStage::ESS_ANY_HIT,
+                    bool(flags & FLAGS::NO_NULL_ANY_HIT_SHADERS)))
+                    return false;
+
+                  // https://docs.vulkan.org/spec/latest/chapters/pipelines.html#VUID-VkRayTracingPipelineCreateInfoKHR-flags-03471
+                  if (!isValidShaderIndex(shaderGroup.closestHit, 
+                    hlsl::ShaderStage::ESS_CLOSEST_HIT,
+                    bool(flags & FLAGS::NO_NULL_CLOSEST_HIT_SHADERS)))
+                    return false;
+
+                  if (!isValidShaderIndex(shaderGroup.intersection, 
+                    hlsl::ShaderStage::ESS_INTERSECTION,
+                    false))
+                    return false;
+                }
+
+                for (const auto& shaderGroup : shaderGroups.misses)
+                {
+                  if (!isValidShaderIndex(shaderGroup.index, 
+                    hlsl::ShaderStage::ESS_MISS, 
+                    false))
+                    return false;
+                }
+
+                for (const auto& shaderGroup : shaderGroups.callables)
+                {
+                  if (!isValidShaderIndex(shaderGroup.index, hlsl::ShaderStage::ESS_CALLABLE, false))
+                    return false;
+                }
+                return true;
+              }
+
+          public:
+            inline bool valid() const
+            {
+              return impl_valid([](const SShaderSpecInfo& info)->bool
+              {
+                if (!info.valid())
+                  return false;
+                return false;
+              });
```

Comment:
thats just repeating over and over, I don't think we need callbacks to check shader stages anymore